File: libraries/ndf/ndf1_chxnm.f
      SUBROUTINE NDF1_CHXNM( XNAME, STATUS )
*+
*  Name:
*     NDF1_CHXNM

*  Purpose:
*     Check an NDF extension name.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF1_CHXNM( XNAME, STATUS )

*  Description:
*     The routine checks the name of an NDF extension for standard
*     form. A standard name must be no more than NDF__SZXNM characters
*     long, must begin with an alphabetic character and continue with
*     alphanumeric characters (including underscore) only. If this test
*     fails, then an error is reported and a STATUS value set.
*     Otherwise, the routine returns without action.

*  Arguments:
*     XNAME = CHARACTER * ( * ) (Given)
*        The extension name to be checked.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Check that the extension name is not too long and that it has
*     the standard form. Report an error if appropriate.

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     19-SEP-1989 (RFWS):
*        Original version.
*     26-SEP-1989 (RFWS):
*        Finished prologue.
*     23-NOV-1989 (RFWS):
*        Changed to use the NDF__SZXNM constant.
*     29-JAN-1990 (RFWS):
*        Removed checks on name registration.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ public constants
      INCLUDE 'NDF_ERR'          ! NDF_ error codes

*  Arguments Given:
      CHARACTER * ( * ) XNAME

*  Status:
      INTEGER STATUS             ! Global status

*  External References:
      LOGICAL CHR_ISNAM          ! Whether a string is a standard name
      INTEGER CHR_LEN            ! Significant length of string

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  If the extension name is too long, or does not have the correct
*  standard form, then report an error.
      IF ( ( CHR_LEN( XNAME ) .GT. NDF__SZXNM ) .OR.
     :     ( .NOT. CHR_ISNAM( XNAME ) ) ) THEN
         STATUS = NDF__NSXNM
         CALL MSG_SETC( 'XNAME', XNAME )
         CALL ERR_REP( 'NDF1_CHXNM_NS',
     :   'Non-standard extension name ''^XNAME'' specified ' //
     :   '(possible programming error).', STATUS )
      END IF
       
*  Call error tracing routine and exit.
      IF ( STATUS .NE. SAI__OK ) CALL NDF1_TRACE( 'NDF1_CHXNM', STATUS )

      END

File: libraries/ndf/ndf1Chxnm.c
#include "sae_par.h"
#include "dat_par.h"
#include "ndf1.h"
#include "ndf_err.h"
#include "ndf_ast.h"
#include "mers.h"

void ndf1Chxnm( const char *xname, size_t start, size_t end, int *status ){
/*
*+
*  Name:
*     ndf1Chxnm

*  Purpose:
*     Check an NDF extension name.

*  Synopsis:
*     void ndf1Chxnm( const char *xname, size_t start, size_t end, int *status )

*  Description:
*     This function checks the name of an NDF extension for standard form.
*     A standard name must be no more than NDF__SZXNM characters long, must
*     begin with an alphabetic character and continue with alphanumeric
*     characters (including underscore) only. If this test fails, then an
*     error is reported and a "status" value set. Otherwise, the function
*     returns without action.

*  Parameters:
*     xname
*        Pointer to a null terminated string holding the extension name to
*        be checked.
*     start
*        The zero-based index of the first character to consider in "xname".
*        The whole string is used if "start" > "end".
*     end
*        The zero-based index of the last character to consider in "xname".
*        The whole string is used if "start" > "end".
*     *status
*        The global status.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   size_t f;          /* Index of first character to use */
   size_t l;          /* Index of last character to use */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Find the indices of the first and last non-space character in the
   section of xname that is to be used. */
   astFandl( xname, start, end, &f, &l );

/* Report an error if there is nothing to check. */
   if( f > l ) {
      *status = NDF__NSXNM;
      errRep( " ", "Blank extension name specified "
              "(possible programming error).", status );

/* If the extension name is too long, or does not have the correct
   standard form, then report an error. */
   } else if( l - f + 1 > NDF__SZXNM || !ndf1Isnam( xname, f, l ) ) {
      *status = NDF__NSXNM;
      msgSetc( "XNAME", xname );
      errRep( " ", "Non-standard extension name '^XNAME' specified "
              "(possible programming error).", status );
   }

/* Call error tracing function and exit. */
   if( *status != SAI__OK ) ndf1Trace( "ndf1Chxnm", status );

}

